File: 0001 TwoSum/TwoSum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        // Brute Force
        // for (int i = 0; i < nums.length - 1; i++) {
        //     for (int j = i + 1; j < nums.length; j++) {
        //         if (nums[i] + nums[j] == target) {
        //             return new int[] {i, j};
        //         }
        //     }
        // }
        // return null;

        // Two-pass Hash Table
        // Map<Integer,Integer> map = new HashMap();
        // for (int i = 0; i < nums.length; i++) {
        //     Integer num = target - nums[i];
        //     if (map.get(num) == null) {
        //         map.put(num, i);
        //     }
        // }

        // for (int i = 0; i < nums.length - 1; i++) {
        //     if (map.get(nums[i]) != null) {
        //         return new int[] {i, map.get(nums[i])};
        //     }
        // }
        // return null;

        // One-pass Hash Table
        Map<Integer,Integer> map = new HashMap();
        for (int i = 0; i < nums.length; i++) {
            int num = nums[i];
            if (map.get(num) != null) {
                return new int[] {map.get(num), i};
            }
            map.put(target - num, i);
        }
        return null;
    }
}

File: 0001 TwoSum/TwoSum.php
<?php
class Solution {
/**
 * @param Integer[] $nums
 * @param Integer $target
 * @return Integer[]
 */
    function twoSum($nums, $target) {
        // Brute Force
        // for ($i = 0; $i < count($nums) - 1; $i++)
        //     for ($j = $i + 1; $j < count($nums); $j++)
        //         if ($nums[$i] + $nums[$j] == $target)
        //             return array($i, $j);
        // return null;

        // Two-pass Hash Table
        // $map = array();
        // for ($i = 0; $i < count($nums); $i++)
        //     $map[$target - $nums[$i]] = $i;

        // for ($i = 0; $i < count($nums) - 1; $i++)
        //     if (array_key_exists($nums[$i], $map))
        //         return array($i, $map[$nums[$i]]);
        // return null;

        // One-pass Hash Table
        $map = [];
        for ($i = 0; $i < count($nums); $i++) {
            $num = $nums[$i];
            if (isset($map[$num]))
                return [$map[$num], $i];
            $map[$target -$nums[$i]] = $i;
        }
        return null;
    }
}
?>
